File: chromium/chrome/browser/preferences/android/java/src/org/chromium/chrome/browser/preferences/PrefServiceBridge.java
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.preferences;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import org.chromium.base.ThreadUtils;
import org.chromium.base.annotations.NativeMethods;

/**
 * PrefServiceBridge is a singleton which provides read and write access to native PrefService.
 */
public class PrefServiceBridge {
    // Singleton constructor. Do not call directly unless for testing purpose.
    @VisibleForTesting
    protected PrefServiceBridge() {}

    private static PrefServiceBridge sInstance;

    /**
     * @return The singleton PrefServiceBridge instance.
     */
    public static PrefServiceBridge getInstance() {
        ThreadUtils.assertOnUiThread();
        if (sInstance == null) {
            sInstance = new PrefServiceBridge();
        }
        return sInstance;
    }

    /**
     * @param preference The name of the preference.
     */
    public void clearPref(@NonNull String preference) {
        PrefServiceBridgeJni.get().clearPref(preference);
    }

    /**
     * @param preference The name of the preference.
     * @return Whether the specified preference is enabled.
     */
    public boolean getBoolean(@NonNull String preference) {
        return PrefServiceBridgeJni.get().getBoolean(preference);
    }

    /**
     * @param preference The name of the preference.
     * @param value The value the specified preference will be set to.
     */
    public void setBoolean(@NonNull String preference, boolean value) {
        PrefServiceBridgeJni.get().setBoolean(preference, value);
    }

    /**
     * @param preference The name of the preference.
     * @return value The value of the specified preference.
     */
    public int getInteger(@NonNull String preference) {
        return PrefServiceBridgeJni.get().getInteger(preference);
    }

    /**
     * @param preference The name of the preference.
     * @param value The value the specified preference will be set to.
     */
    public void setInteger(@NonNull String preference, int value) {
        PrefServiceBridgeJni.get().setInteger(preference, value);
    }

    /**
     * @param preference The name of the preference.
     * @return value The value of the specified preference.
     */
    @NonNull
    public String getString(@NonNull String preference) {
        return PrefServiceBridgeJni.get().getString(preference);
    }

    /**
     * @param preference The name of the preference.
     * @param value The value the specified preference will be set to.
     */
    public void setString(@NonNull String preference, @NonNull String value) {
        PrefServiceBridgeJni.get().setString(preference, value);
    }

    /**
     * @param preference The name of the preference.
     * @return Whether the specified preference is managed.
     */
    public boolean isManagedPreference(@NonNull String preference) {
        return PrefServiceBridgeJni.get().isManagedPreference(preference);
    }

    @VisibleForTesting
    public static void setInstanceForTesting(@Nullable PrefServiceBridge instanceForTesting) {
        sInstance = instanceForTesting;
    }

    @NativeMethods
    interface Natives {
        void clearPref(String preference);
        boolean getBoolean(String preference);
        void setBoolean(String preference, boolean value);
        int getInteger(String preference);
        void setInteger(String preference, int value);
        String getString(String preference);
        void setString(String preference, String value);
        boolean isManagedPreference(String preference);
    }
}

File: chromium/chrome/browser/android/preferences/pref_service_bridge.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <jni.h>

#include <string>

#include "base/android/jni_string.h"
#include "base/android/scoped_java_ref.h"
#include "chrome/browser/preferences/jni_headers/PrefServiceBridge_jni.h"
#include "chrome/browser/profiles/profile_manager.h"
#include "components/prefs/pref_service.h"

namespace {

using base::android::JavaParamRef;

PrefService* GetPrefService() {
  return ProfileManager::GetActiveUserProfile()
      ->GetOriginalProfile()
      ->GetPrefs();
}

}  // namespace

// ----------------------------------------------------------------------------
// Native JNI methods
// ----------------------------------------------------------------------------

static void JNI_PrefServiceBridge_ClearPref(
    JNIEnv* env,
    const JavaParamRef<jstring>& j_preference) {
  GetPrefService()->ClearPref(
      base::android::ConvertJavaStringToUTF8(env, j_preference));
}

static jboolean JNI_PrefServiceBridge_GetBoolean(
    JNIEnv* env,
    const JavaParamRef<jstring>& j_preference) {
  return GetPrefService()->GetBoolean(
      base::android::ConvertJavaStringToUTF8(env, j_preference));
}

static void JNI_PrefServiceBridge_SetBoolean(
    JNIEnv* env,
    const JavaParamRef<jstring>& j_preference,
    const jboolean j_value) {
  GetPrefService()->SetBoolean(
      base::android::ConvertJavaStringToUTF8(env, j_preference), j_value);
}

static jint JNI_PrefServiceBridge_GetInteger(
    JNIEnv* env,
    const JavaParamRef<jstring>& j_preference) {
  return GetPrefService()->GetInteger(
      base::android::ConvertJavaStringToUTF8(env, j_preference));
}

static void JNI_PrefServiceBridge_SetInteger(
    JNIEnv* env,
    const JavaParamRef<jstring>& j_preference,
    const jint j_value) {
  GetPrefService()->SetInteger(
      base::android::ConvertJavaStringToUTF8(env, j_preference), j_value);
}

static base::android::ScopedJavaLocalRef<jstring>
JNI_PrefServiceBridge_GetString(JNIEnv* env,
                                const JavaParamRef<jstring>& j_preference) {
  return base::android::ConvertUTF8ToJavaString(
      env, GetPrefService()->GetString(
               base::android::ConvertJavaStringToUTF8(env, j_preference)));
}

static void JNI_PrefServiceBridge_SetString(
    JNIEnv* env,
    const JavaParamRef<jstring>& j_preference,
    const base::android::JavaParamRef<jstring>& j_value) {
  GetPrefService()->SetString(
      base::android::ConvertJavaStringToUTF8(env, j_preference),
      base::android::ConvertJavaStringToUTF8(env, j_value));
}

static jboolean JNI_PrefServiceBridge_IsManagedPreference(
    JNIEnv* env,
    const JavaParamRef<jstring>& j_preference) {
  return GetPrefService()->IsManagedPreference(
      base::android::ConvertJavaStringToUTF8(env, j_preference));
}
